File: landmask.py
"""
landmask.py — Détection terre/mer pour le fallback Python (routing.py).

Utilise le même dataset que le moteur Rust (engine/data/coastline.geojson),
indexé via shapely.STRtree pour des requêtes point-in-polygon rapides.
Le module charge le GeoJSON une seule fois (lazy, à la 1re utilisation).
"""

from __future__ import annotations

import json
import logging
from functools import lru_cache
from pathlib import Path
from typing import List

from shapely.geometry import LineString, MultiPolygon, Point, Polygon, shape
from shapely.strtree import STRtree

logger = logging.getLogger(__name__)

_COASTLINE_PATH = (
    Path(__file__).resolve().parent / "engine" / "data" / "coastline.geojson"
)


@lru_cache(maxsize=1)
def _load_index() -> tuple[STRtree, list]:
    """Charge le GeoJSON, retourne (STRtree, liste de polygones)."""
    if not _COASTLINE_PATH.exists():
        logger.warning("landmask: coastline.geojson introuvable, fallback désactivé")
        return STRtree([]), []

    with _COASTLINE_PATH.open("r", encoding="utf-8") as f:
        gj = json.load(f)

    polys: List[Polygon | MultiPolygon] = []
    for feat in gj.get("features", []):
        geom = feat.get("geometry")
        if not geom:
            continue
        try:
            g = shape(geom)
        except Exception:
            continue
        if g.is_empty:
            continue
        if g.geom_type in ("Polygon", "MultiPolygon"):
            polys.append(g)

    tree = STRtree(polys)
    logger.info("landmask: %d polygones chargés", len(polys))
    return tree, polys


def is_on_land(lat: float, lon: float) -> bool:
    """True si le point (lat, lon) est sur une masse terrestre."""
    tree, polys = _load_index()
    if not polys:
        return False
    pt = Point(lon, lat)
    # STRtree.query renvoie les indices des candidats dont la bbox intersecte
    candidate_idx = tree.query(pt)
    for i in candidate_idx:
        if polys[i].contains(pt):
            return True
    return False


def crosses_land_buffered(
    lat1: float,
    lon1: float,
    lat2: float,
    lon2: float,
    buffer_nm: float = 0.0,
) -> bool:
    """
    True si le segment (lat1,lon1)→(lat2,lon2) traverse une masse terrestre,
    en ignorant un buffer de `buffer_nm` autour de chaque extrémité.

    Le buffer permet le départ/arrivée depuis un mouillage côtier (l'endpoint
    peut être marqué "on land" selon la résolution du dataset Natural Earth 10m).
    """
    tree, polys = _load_index()
    if not polys:
        return False

    # Conversion approximative degrés↔NM (1° lat ≈ 60 NM, lon ajusté par cos(lat))
    import math

    dlat = lat2 - lat1
    dlon = (lon2 - lon1) * math.cos(math.radians(lat1))
    approx_nm = math.sqrt(dlat * dlat + dlon * dlon) * 60.0
    if approx_nm <= 0.0:
        return False
    if 2 * buffer_nm >= approx_nm:
        # Segment trop court par rapport au buffer : on considère qu'il ne traverse pas.
        return False

    # On échantillonne le segment, en ignorant les zones tampon autour des extrémités.
    n = max(2, min(60, int(math.ceil(approx_nm * 2.0))))
    t_buf = buffer_nm / approx_nm
    line = LineString([(lon1, lat1), (lon2, lat2)])
    candidate_idx = tree.query(line)
    if len(candidate_idx) == 0:
        return False

    candidates = [polys[i] for i in candidate_idx]
    for i in range(1, n):
        t = i / n
        if t < t_buf or t > 1.0 - t_buf:
            continue
        lat = lat1 + dlat * t
        lon = lon1 + (lon2 - lon1) * t
        pt = Point(lon, lat)
        for poly in candidates:
            if poly.contains(pt):
                return True
    return False


def warm_up() -> None:
    """Force le chargement de l'index (utile au démarrage worker)."""
    _load_index()

File: engine/src/landmask.rs
//! Détection terre/mer basée sur Natural Earth 10m (land + minor_islands).
//!
//! Au premier appel : parse le GeoJSON embarqué, construit un R-tree des bounding boxes
//! pour pré-filtrer rapidement, puis test point-in-polygon sur les candidats.

use geocrate::{BoundingRect, Contains, Coord, LineString, MultiPolygon, Point, Polygon};
use geojson::{GeoJson, Geometry, Value};
use once_cell::sync::Lazy;
use rstar::{RTree, RTreeObject, AABB};

const COASTLINE_GEOJSON: &str = include_str!("../data/coastline.geojson");

struct IndexedPoly {
    poly: MultiPolygon<f64>,
    bbox: AABB<[f64; 2]>,
}

impl RTreeObject for IndexedPoly {
    type Envelope = AABB<[f64; 2]>;
    fn envelope(&self) -> Self::Envelope {
        self.bbox
    }
}

fn ring_to_linestring(ring: Vec<Vec<f64>>) -> LineString<f64> {
    let coords: Vec<Coord<f64>> = ring
        .into_iter()
        .filter_map(|c| {
            if c.len() >= 2 {
                Some(Coord { x: c[0], y: c[1] })
            } else {
                None
            }
        })
        .collect();
    LineString::new(coords)
}

fn polygon_from_rings(rings: Vec<Vec<Vec<f64>>>) -> Option<Polygon<f64>> {
    let mut iter = rings.into_iter();
    let exterior = ring_to_linestring(iter.next()?);
    let interiors: Vec<LineString<f64>> = iter.map(ring_to_linestring).collect();
    Some(Polygon::new(exterior, interiors))
}

fn collect_polygons(geom: &Geometry, out: &mut Vec<MultiPolygon<f64>>) {
    match &geom.value {
        Value::Polygon(rings) => {
            if let Some(poly) = polygon_from_rings(rings.clone()) {
                out.push(MultiPolygon(vec![poly]));
            }
        }
        Value::MultiPolygon(mps) => {
            let polys: Vec<Polygon<f64>> = mps
                .iter()
                .filter_map(|r| polygon_from_rings(r.clone()))
                .collect();
            if !polys.is_empty() {
                out.push(MultiPolygon(polys));
            }
        }
        Value::GeometryCollection(gs) => {
            for g in gs {
                collect_polygons(g, out);
            }
        }
        _ => {}
    }
}

fn build_index() -> RTree<IndexedPoly> {
    let parsed: GeoJson = COASTLINE_GEOJSON
        .parse()
        .expect("coastline.geojson invalide");
    let fc = match parsed {
        GeoJson::FeatureCollection(fc) => fc,
        _ => panic!("coastline.geojson : FeatureCollection attendue"),
    };

    let mut polys: Vec<MultiPolygon<f64>> = Vec::new();
    for feat in fc.features {
        if let Some(geom) = feat.geometry {
            collect_polygons(&geom, &mut polys);
        }
    }

    let indexed: Vec<IndexedPoly> = polys
        .into_iter()
        .filter_map(|mp| {
            let r = mp.bounding_rect()?;
            let bbox = AABB::from_corners([r.min().x, r.min().y], [r.max().x, r.max().y]);
            Some(IndexedPoly { poly: mp, bbox })
        })
        .collect();

    RTree::bulk_load(indexed)
}

static INDEX: Lazy<RTree<IndexedPoly>> = Lazy::new(build_index);

/// True si le point (lat, lon) est sur une masse terrestre.
pub fn is_on_land(lat: f64, lon: f64) -> bool {
    let p = Point::new(lon, lat);
    let envelope = AABB::from_point([lon, lat]);
    INDEX
        .locate_in_envelope_intersecting(&envelope)
        .any(|ip| ip.poly.contains(&p))
}

/// Force la construction de l'index (utile pour mesurer ou pré-charger).
pub fn warm_up() {
    Lazy::force(&INDEX);
}

/// True si le segment lat1,lon1 → lat2,lon2 traverse une masse terrestre,
/// en ignorant un "buffer" de `buffer_nm` autour de chaque extrémité.
/// Le buffer permet le départ/arrivée depuis un mouillage côtier (endpoint
/// peut être lui-même on-land selon la résolution du dataset).
pub fn crosses_land_buffered(
    lat1: f64,
    lon1: f64,
    lat2: f64,
    lon2: f64,
    buffer_nm: f64,
) -> bool {
    let dlat = lat2 - lat1;
    let dlon = (lon2 - lon1) * lat1.to_radians().cos();
    let approx_nm = (dlat * dlat + dlon * dlon).sqrt() * 60.0;
    let usable_nm = approx_nm - 2.0 * buffer_nm;
    if usable_nm < 0.2 {
        return false;
    }
    let n = ((approx_nm * 2.0).ceil() as usize).clamp(2, 60);
    let t_buf = if approx_nm > 0.0 {
        buffer_nm / approx_nm
    } else {
        0.0
    };
    for i in 1..n {
        let t = i as f64 / n as f64;
        if t < t_buf || t > 1.0 - t_buf {
            continue;
        }
        let lat = lat1 + (lat2 - lat1) * t;
        let lon = lon1 + (lon2 - lon1) * t;
        if is_on_land(lat, lon) {
            return true;
        }
    }
    false
}

/// True si le segment lat1,lon1 → lat2,lon2 traverse une masse terrestre.
/// Wrapper sans buffer, conservé pour rétro-compat.
pub fn crosses_land(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> bool {
    crosses_land_buffered(lat1, lon1, lat2, lon2, 0.0)
}
